src/common/log.cpp

```diff
@@ -874,17 +880,42 @@ void wxLogStderr::DoLogText(const wxString& msg)
 
 #if wxUSE_STD_IOSTREAM
 #include "wx/ioswrap.h"
-wxLogStream::wxLogStream(wxSTD ostream *ostr)
+wxLogStream::wxLogStream(wxSTD ostream *ostr, const wxMBConv &conv):
+m_conv(conv.Clone())
 {
     if ( ostr == NULL )
         m_ostr = &wxSTD cerr;
     else
         m_ostr = ostr;
 }
 
+wxLogStream::~wxLogStream()
+{
+    delete m_conv;
+}
+
 void wxLogStream::DoLogText(const wxString& msg)
 {
-    (*m_ostr) << msg << wxSTD endl;
+    wxString msgLF(msg);
+    if ( msgLF.empty() || *msgLF.rbegin() != '\n' )
+        msgLF += '\n';
+
+    int lengthAdjust = 0;
+
+#if defined(__WINDOWS__)
+    // Determine whether the encoding is UTF-16. In that case, the file
+    // should have been opened in "wb" mode, and EOL-style must be handled
+    // here. Also one bytes less should be written.
```

Comment:
Could you please explain why is this so?

It seems to me that writing 2 trailing NUL bytes should work just fine too, so while it's probably not harmful to write one byte less, I don't see why would it be necessary, could you please explain?

P.S. Please don't change the code, I've started merging it before thinking about this question and I've already done some changes on top of this version.

Reply:
I don't remember if I figured out the exact 'why', but what happens when not subtracting `1` from `buf.length()` is:
* The number of bytes written to the file is non-even, which is strange for a file which is supposed to multiple of 2 bytes.
* Notepad in Windows 7 doesn't recognize such files as Unicode, and cannot properly open them.

Reply:
This is really weird. Maybe we should just subtract `GetMBNulLen()` then? Because I just don't understand how can it be correct to write a half of UTF-16 terminating NUL, either it should be written entirely or not at all...

Reply:
It cannot be correct, and indeed it seems to me that subtracting `1` *prevents* writing half of a UTF-16 terminating NUL. In other words, when the file is written correctly, there is no terminating NUL of any kind.

A key question here is why does `cWX2MB` return a `buf` whose length is not even, when converting to UTF-16.

Reply:
Have you already checked/confirmed that `cWX2MB()` indeed returns an odd-sized buffer in this case? This looks like such a bad bug to me that I didn't even think we could have it :-( If we do, it definitely needs to be fixed, but I'm working on something else right now, so I won't do it immediately.

Reply:
I did do some testing, and it looks like the buffer size is odd. I created a PR to demonstrate this in a unittest: https://github.com/wxWidgets/wxWidgets/pull/569

Reply:
Thanks, I'll try to look at this later, this really seems very wrong, but I'm not sure what the consequences of changing it could be.

Reply:
I've opened #579 to fix, the hack with length adjustment shouldn't be necessary any more once it's merged, could you please update this PR when it's done? TIA!

Reply:
Just FYI, #579 is finally merged.

Reply:
I have rebased this PR to current master and made the required modifications (i.e. removing the whole `lengthAdjust`.)